### PR TITLE
add support for proxy

### PR DESCRIPTION
**Pull request by hrittikhere**
**Description:** <Describe what has changed. 
Adding a feature to support proxy configuration field in all exporters

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector/issues/5761

**Testing:** < Describe what testing was performed and which tests were added.>

**Documentation:** < Describe the documentation added.>

### CODE REVIEW

File: config/confighttp/confighttp.go [hunk 1]
```diff
@@ -32,6 +33,9 @@ type HTTPClientSettings struct {
 	// The target URL to send data to (e.g.: http://some.url:9411/v1/traces).
 	Endpoint string `mapstructure:"endpoint"`
 
+	// Proxy setting for the collector
+	Proxy string `mapstructure:"proxy"`
```
Comment (jpkrohling) [inline config/confighttp/confighttp.go:37]: perhaps ProxyURL?

```suggestion
	ProxyURL string `mapstructure:"proxy_url"`
```

File: config/confighttp/confighttp.go [hunk 2]
```diff
@@ -124,6 +128,16 @@ func (hcs *HTTPClientSettings) ToClient(host component.Host, settings component.
...
+	// Setting the Proxy URL
+	if hcs.Proxy != "" {
+		proxyURL, err := url.Parse(hcs.Proxy)
+		if err != nil {
+			return nil, err
+		} else {
+			transport.Proxy = http.ProxyURL(proxyURL)
+		}
+	}
+
```
Comment (jpkrohling) [inline config/confighttp/confighttp.go:140]: Looks good, but needs a new test.

Comment (hrittikhere) [reply config/confighttp/confighttp.go:140]: Yes, I am working on the tests but facing a bit of issue configuring parameters. I'll ping you as soon as I have them :)

File: config/confighttp/confighttp.go [hunk 3]
```diff
@@ -124,6 +128,16 @@ func (hcs *HTTPClientSettings) ToClient(host component.Host, settings component.
 		transport.IdleConnTimeout = *hcs.IdleConnTimeout
 	}
 
+	// Setting the Proxy URL
+	if hcs.Proxy != "" {
+		proxyURL, err := url.Parse(hcs.Proxy)
+		if err != nil {
+			return nil, err
+		} else {
```
Comment (jkandasa) [inline config/confighttp/confighttp.go:136]: we can omit this else

Comment (hrittikhere) [reply config/confighttp/confighttp.go:136]:
```go 
else {			
    transport.Proxy = http.ProxyURL(proxyURL)
		}
```

like this whole part?

Comment (jkandasa) [reply config/confighttp/confighttp.go:136]: @hrittikhere no,
```go
if hcs.Proxy != "" {
    proxyURL, err := url.Parse(hcs.Proxy)
    if err != nil {
        return nil, err
    } 
    transport.Proxy = http.ProxyURL(proxyURL)
}
```

Comment (hrittikhere) [reply config/confighttp/confighttp.go]: Fixed ✅

**Comment by jpkrohling** [comment]
@hrittikhere, do you have an status update regarding this one?

**Comment by jpkrohling** [comment]
@hrittikhere, are you able to fix the merge conflicts?

**Comment by hrittikhere** [comment]
@jpkrohling does this looks good?

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
LGTM!

**Comment by jpkrohling** [comment]
Apparently, there's still a linting failure:

```
make[2]: Entering directory '/home/runner/work/opentelemetry-collector/opentelemetry-collector/config/confighttp'
/home/runner/work/opentelemetry-collector/opentelemetry-collector/.tools/golangci-lint run
confighttp.go:147: File is not `gofmt`-ed with `-s` (gofmt)
make[2]: *** [../../Makefile.Common:74: lint] Error 1
make[1]: *** [Makefile:123: config/confighttp] Error 2
make[1]: *** Waiting for unfinished jobs....
  
Error: confighttp.go:1[41](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/6330517394/job/17424908838?pr=8339#step:5:42):13: shadow: declaration of "err" shadows declaration at line 108 (govet)
		proxyURL, err := url.Parse(hcs.ProxyURL)
```

File: config/confighttp/confighttp.go [hunk 4]
```diff
@@ -132,6 +136,15 @@ func (hcs *HTTPClientSettings) ToClient(host component.Host, settings component.
 		transport.IdleConnTimeout = *hcs.IdleConnTimeout
 	}
 
+	// Setting the Proxy URL
+	if hcs.ProxyURL != "" {
+		proxyURL, err := url.Parse(hcs.ProxyURL)
```
Comment (astencel-sumo) [inline config/confighttp/confighttp.go:141]: Shouldn't this be rather [url.ParseRequestURI](https://pkg.go.dev/net/url#ParseRequestURI)?
```suggestion
		proxyURL, err := url.ParseRequestURI(hcs.ProxyURL)
```

**Comment by astencel-sumo** [comment]
Changelog entry is missing right? https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#adding-a-changelog-entry

**Comment by hrittikhere** [comment]
Sure I'll add the changelog entry over the weekend. Sorry I missed the comments :(

**Comment by hrittikhere** [comment]
I tried fixing the lint and I think it works. Maybe we can re run? 

```
[build-and-test/lint             ]   ✅  Success - Main goimpi
[build-and-test/lint             ] ⭐ Run Post Cache Go
[build-and-test/lint             ]   🐳  docker exec cmd=[node /var/run/act/actions/actions-cache@v3/dist/save/index.js] user= workdir=
[build-and-test/lint             ]   💬  ::debug::Cache state/key: go-cache-linux-54ab5a16ba9fdf26f1dc2d13615853606f75f614eb63f91ef59afcc63707bb55
| Cache hit occurred on the primary key go-cache-Linux-54ab5a16ba9fdf26f1dc2d13615853606f75f614eb63f91ef59afcc63707bb55, not saving cache.
[build-and-test/lint             ]   ✅  Success - Post Cache Go
[build-and-test/lint             ] ⭐ Run Post Setup Go
[build-and-test/lint             ]   🐳  docker exec cmd=[node /var/run/act/actions/actions-setup-go@v4/dist/cache-save/index.js] user= workdir=
| [command]/opt/hostedtoolcache/go/1.20.10/x64/bin/go env GOMODCACHE
| [command]/opt/hostedtoolcache/go/1.20.10/x64/bin/go env GOCACHE
| /root/go/pkg/mod
| /root/.cache/go-build
[build-and-test/lint             ]   💬  ::debug::Checking zstd --quiet --version
[build-and-test/lint             ]   💬  ::debug::1.4.8
[build-and-test/lint             ]   💬  ::debug::zstd version: 1.4.8
[build-and-test/lint             ]   💬  ::debug::implicitDescendants 'false'
[build-and-test/lint             ]   💬  ::debug::followSymbolicLinks 'true'
[build-and-test/lint             ]   💬  ::debug::implicitDescendants 'false'
[build-and-test/lint             ]   💬  ::debug::omitBrokenSymbolicLinks 'true'
[build-and-test/lint             ]   💬  ::debug::Search path '/root/go/pkg/mod'
[build-and-test/lint             ]   💬  ::debug::Search path '/root/.cache/go-build'
[build-and-test/lint             ]   💬  ::debug::Matched: ../../root/go/pkg/mod
[build-and-test/lint             ]   💬  ::debug::Matched: ../../root/.cache/go-build
[build-and-test/lint             ]   💬  ::debug::Cache Paths:
[build-and-test/lint             ]   💬  ::debug::["../../root/go/pkg/mod","../../root/.cache/go-build"]
[build-and-test/lint             ]   💬  ::debug::Archive Path: /tmp/236e3b8b-ebb9-41b0-b8ee-c3d54308cc3d/cache.tzst
| [command]/usr/bin/tar --posix -cf cache.tzst --exclude cache.tzst -P -C /workspaces/opentelemetry-collector --files-from manifest.txt --use-compress-program zstdmt
[build-and-test/lint             ]   💬  ::debug::File Size: 357739713
[build-and-test/lint             ]   💬  ::debug::Reserving Cache
[build-and-test/lint             ]   💬  ::debug::Resource Url: http://172.16.5.4:43717/_apis/artifactcache/caches
ERRO[0486] POST /_apis/artifactcache/caches: already exist  module=artifactcache
[build-and-test/lint             ]   🚧  ::warning::Failed to save: {"error":"already exist"}
[build-and-test/lint             ]   ✅  Success - Post Setup Go
[build-and-test/lint             ] 🏁  Job succeeded
```

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Sure, build started!

**Comment by hrittikhere** [comment]
I think we're good to go?  The `build-and-test / test-coverage (pull_request) Failing after 6m` fails, but we have tests in place? :)

**Comment by jpkrohling** [comment]
We just need an @open-telemetry/collector-approvers to review and push the button!

**Review by astencel-sumo — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]

File: .chloggen/proxy.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: exporter
```
Comment (codeboten) [inline .chloggen/proxy.yaml:7]:
```suggestion
component: confighttp
```

File: .chloggen/proxy.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Support proxy configuration field in all exporters
```
Comment (codeboten) [inline .chloggen/proxy.yaml:10]:
```suggestion
note: Support proxy configuration field in all exporters that support confighttp
```

File: config/confighttp/confighttp_test.go [hunk 7]
```diff
@@ -208,6 +208,48 @@ func TestDefaultHTTPClientSettings(t *testing.T) {
...
+	}{
+		{
+			desc:        "default config",
+			expectedURL: nil,
+		},
+		{
+			desc:        "proxy is set",
+			proxyURL:    "http://proxy.example.com:8080",
+			expectedURL: &url.URL{Scheme: "http", Host: "proxy.example.com:8080"},
+		},
```
Comment (dmitryax) [inline config/confighttp/confighttp_test.go:225]: Can you please add a case with an invalid proxyURL string to get 100% test coverage?

Comment (hrittikhere) [reply config/confighttp/confighttp_test.go:225]: @dmitryax I am travelling till the Wed without a laptop so it will be bit hard.

**Comment by hrittikhere** [comment]
Can we run the test coverage once more? :)

**Review by dmitryax — APPROVED** [review]

**Comment by hrittikhere** [comment]
Thanks @jpkrohling for the help throughout! Couldn't have done it without you :)

**Pull request merged**